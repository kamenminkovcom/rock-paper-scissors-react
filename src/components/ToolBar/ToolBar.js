import React from 'react';
import HandButton from './HandButton/HandButton';
import PropTypes from 'prop-types';

const ToolBar = ({handTypes, choseHandType}) => (
    <div className="tool-bar">
        {Object.values(handTypes).map(h => (
            <HandButton handType={h}
                        handleClick={() => choseHandType(h)}
                        key={h}/>
        ))}
    </div>
);

ToolBar.propTypes = {
    handTypes: PropTypes.object.isRequired,
    choseHandType: PropTypes.func.isRequired
};

export default ToolBar;