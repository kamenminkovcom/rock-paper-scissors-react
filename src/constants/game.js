export const handTypes = {
    paper: 'paper',
    rock: 'rock',
    scissors: 'scissors',
};

export const gameRules = {
    [handTypes.paper]: {
        [handTypes.paper]: 0,
        [handTypes.rock]: 1,
        [handTypes.scissors]: 0,
    },
    [handTypes.rock]: {
        [handTypes.rock]: 0,
        [handTypes.scissors]: 1,
        [handTypes.paper]: 0
    },
    [handTypes.scissors]: {
        [handTypes.scissors]: 0,
        [handTypes.paper]: 1,
        [handTypes.rock]: 0
    }
};

export const possibleResults = {
    win: 'win',
    lose: 'lose',
    draw: 'draw'
};

//may we would like to change classNames and it should not break the logic
export const classNames = {
    [possibleResults.win]: 'win',
    [possibleResults.lose]: 'lose',
    [possibleResults.draw]: 'draw'
};

export const gameHistoryCount = 5;

export const defaultChoices = {
    cpuChoice: undefined,
    playerChoice: undefined
};