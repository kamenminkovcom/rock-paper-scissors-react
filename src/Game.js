import React from 'react';
import Hand from './components/Hand/Hand';
import Desk from './components/Desk/Desk';
import ScoreBar from './components/ScoreBar/ScoreBar';
import ToolBar from './components/ToolBar/ToolBar';
import {bindTo, randomizeHandTypes} from './utils/utils';
import {handTypes, gameRules} from './constants/game';

class Game extends React.Component {
    constructor(props) {
        super(props);
        bindTo('startGame', 'choseHandType')(this);
        this.initialState = {
            cpuChoice: null,
            playerChoice: null,
            cpuScore: 0,
            playerScore: 0
        };
        this.state = {
            isPlaying: false,
            cpuScore: this.initialState.cpuScore,
            playerScore: this.initialState.playerScore
        };
    }

    startGame() {
        this.setState(prevState => ({
            isPlaying: !prevState.isPlaying,
            cpuChoice: this.initialState.cpuChoice,
            playerChoice: this.initialState.cpuChoice
        }));
    }

    choseHandType(playerChoice) {
        const handTypesCollection = Object.keys(handTypes);
        const handTypesRandomIndex = randomizeHandTypes(handTypesCollection.length);
        const cpuChoice = handTypesCollection[handTypesRandomIndex];
        //Results from the current game
        const cpuPoints = gameRules[cpuChoice][playerChoice];
        const playerPoints = gameRules[playerChoice][cpuChoice];
        this.setState(prevState => ({
            cpuChoice,
            playerChoice,
            isPlaying: !prevState.isPlaying,
            cpuScore: prevState.cpuScore += cpuPoints,
            playerScore: prevState.playerScore += playerPoints
        }))
    }

    render() {
        const {
            isPlaying,
            cpuChoice,
            playerChoice,
            cpuScore,
            playerScore
        } = this.state;

        return (
            <div>
                <ScoreBar cpuScores={cpuScore} playerScores={playerScore}/>
                <div className="game-container">
                    <Hand isCpu={!!cpuChoice} handType={cpuChoice ? cpuChoice : 'rules'}/>
                    <Desk playFunc={this.startGame}
                          isPlaying={isPlaying}
                          history={[]}  />
                    <Hand handType={playerChoice ? playerChoice : 'rules'}/>
                </div>
                {isPlaying && <ToolBar choseHandType={this.choseHandType}
                                       handTypes={handTypes}/>}
            </div>
        )
    }
}

export default Game;