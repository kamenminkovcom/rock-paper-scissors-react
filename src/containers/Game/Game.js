import React from 'react';
import uid from 'uid';
import Hand from '../../components/Hand/Hand';
import Desk from '../../components/Desk/Desk';
import ScoreBar from '../../components/ScoreBar/ScoreBar';
import ToolBar from '../../components/ToolBar/ToolBar';
import {bindTo, randomizeHandTypes, getResult, resizeCollection} from '../../utils/utils';
import {handTypes, gameRules, gameHistoryCount, defaultChoice} from '../../constants/game';

import './Game.css';

class Game extends React.Component {
    constructor(props) {
        super(props);
        bindTo('startGame', 'choseHandType')(this);
        this.state = {
            isPlaying: false,
            cpuScore: 0,
            playerScore: 0,
            history: []
        };
    }

    startGame() {
        this.setState(prevState => ({
            isPlaying: !prevState.isPlaying,
            cpuChoice: defaultChoice,
            playerChoice: defaultChoice
        }));
    }

    choseHandType(playerChoice) {
        const handTypesCollection = Object.values(handTypes);
        const handTypesRandomIndex = randomizeHandTypes(handTypesCollection.length);
        const cpuChoice = handTypesCollection[handTypesRandomIndex];
        //Results from the current game
        const cpuPoints = gameRules[cpuChoice][playerChoice];
        const playerPoints = gameRules[playerChoice][cpuChoice];
        const gameResult = {
            cpuChoice,
            playerChoice,
            result: getResult(cpuPoints, playerPoints),
            id: uid() //unique id for every array element, we are going to use it as key
        };
        this.setState(prevState => ({
            cpuChoice,
            playerChoice,
            isPlaying: !prevState.isPlaying,
            cpuScore: prevState.cpuScore += cpuPoints,
            playerScore: prevState.playerScore += playerPoints,
            history: resizeCollection([gameResult, ...prevState.history], gameHistoryCount)
        }))
    }

    render() {
        const {
            isPlaying,
            cpuChoice,
            playerChoice,
            cpuScore,
            playerScore,
            history
        } = this.state;

        return (
            <div>
                <ScoreBar cpuScores={cpuScore} playerScores={playerScore}/>
                <div className="game-container">
                    <Hand isCpu={!!cpuChoice} handType={cpuChoice ? cpuChoice : 'rules'}/>
                    <Desk playFunc={this.startGame}
                          isPlaying={isPlaying}
                          history={history}/>
                    <Hand handType={playerChoice ? playerChoice : 'rules'}/>
                </div>
                {isPlaying && <ToolBar choseHandType={this.choseHandType}
                                       handTypes={handTypes}/>}
            </div>
        )
    }
}

export default Game;