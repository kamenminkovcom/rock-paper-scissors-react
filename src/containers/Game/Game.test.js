import React from 'react';
import {shallow, mount} from 'enzyme';
import Game from './Game';
import ToolBar from '../../components/ToolBar/ToolBar';
import {handTypes, gameHistoryCount, defaultChoice} from '../../constants/game';


describe('<Game/>', () => {
    let wrapper;
    const defaultChoiceType = typeof defaultChoice;

    beforeEach(() => {
        wrapper = shallow(<Game/>);
    });

    it('Should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Should initialize history as array', () => {
        expect(Array.isArray(wrapper.state().history)).toEqual(true);
    });

    it('Should initialize empty history array', () => {
        expect(wrapper.state().history.length === 0).toEqual(true);
    });

    it('Should push at history after player plays', () => {
        wrapper.instance().choseHandType(handTypes.rock);
        expect(wrapper.state().history.length === 1).toEqual(true);
    });

    it('Should push at history after multiple plays', () => {
        const times = 4;
        for (let i = 1; i <= times; i++) {
            wrapper.instance().choseHandType(handTypes.rock);
        }
        expect(wrapper.state().history.length === times).toEqual(true);
    });

    it('History should not be more than gameHistoryCount constant', () => {
        const times = gameHistoryCount + 1;
        for (let i = 1; i <= times; i++) {
            wrapper.instance().choseHandType(handTypes.rock);
        }
        expect(wrapper.state().history.length === gameHistoryCount).toEqual(true);
    });

    it('Should initialize cpuChoice correctly', () => {
        expect(typeof wrapper.state().cpuChoice === defaultChoiceType).toEqual(true);
    });

    it('Should initialize playerChoice correctly', () => {
        expect(typeof wrapper.state().playerChoice === defaultChoiceType).toEqual(true);
    });

    it('Should set playerChoice after play', () => {
        wrapper.instance().choseHandType(handTypes.rock);
        expect(typeof wrapper.state().playerChoice === 'string').toEqual(true);
    });

    it('Should set cpuChoice after play', () => {
        wrapper.instance().choseHandType(handTypes.rock);
        expect(typeof wrapper.state().cpuChoice === 'string').toEqual(true);
    });

    it('Should set correct playerChoice after play', () => {
        wrapper.instance().choseHandType(handTypes.rock);
        expect(wrapper.state().playerChoice === handTypes.rock).toEqual(true);
    });

    it('Should set correct one of types cpuChoice', () => {
        wrapper.instance().choseHandType(handTypes.rock);
        expect(Object.values(handTypes).includes(wrapper.state().cpuChoice)).toEqual(true);
    });

    it('Should initialize isPlaying correctly', () => {
       expect(wrapper.state().isPlaying).toEqual(false);
    });

    it('Should set inPlaying correctly after play buton is clicked', () => {
        wrapper.instance().startGame();
        expect(wrapper.state().isPlaying).toEqual(true);
    });

    it('Should set default cpuChoice after play button is clicked for second time', () => {
        wrapper.instance().startGame();
        wrapper.instance().choseHandType(handTypes.rock);
        wrapper.instance().startGame();
        expect(typeof wrapper.state().cpuChoice === defaultChoiceType).toEqual(true);
    });

    it('Should set default playerChoice after play button is clicked for second time', () => {
        wrapper.instance().startGame();
        wrapper.instance().choseHandType(handTypes.rock);
        wrapper.instance().startGame();
        expect(typeof wrapper.state().playerChoice === defaultChoiceType).toEqual(true);
    });

    it('Should initialize correctly playerScore', () => {
       expect(wrapper.state().playerScore === 0).toEqual(true);
    });

    it('Should initialize correctly cpuScore', () => {
        expect(wrapper.state().cpuScore === 0).toEqual(true);
    });

    it('Should not render ToolBar initially', () => {
        const mountedWrapper = mount(<Game/>);
        expect(mountedWrapper.find(ToolBar).children().length === 0).toEqual(true);
    });

    it('Should not render ToolBar if choseHandType is called', () => {
        const mountedWrapper = mount(<Game/>);
        mountedWrapper.instance().startGame();
        mountedWrapper.instance().choseHandType(handTypes.paper);
        expect(mountedWrapper.find(ToolBar).children().length === 0).toEqual(true);
    });

    it('Should render ToolBar if isPlaying is true', () => {
        const mountedWrapper = mount(<Game/>);
        mountedWrapper.setState({isPlaying: true});
        expect(mountedWrapper.find(ToolBar).children().length > 0).toEqual(true);
    });
});