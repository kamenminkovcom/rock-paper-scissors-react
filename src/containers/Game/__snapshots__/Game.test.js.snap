// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game/> Should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Game />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ScoreBar
          cpuScores={0}
          playerScores={0}
/>,
        <div
          className="game-container"
>
          <Hand
                    handType="rules"
                    isCpu={false}
          />
          <Desk
                    history={Array []}
                    isPlaying={false}
                    playFunc={[Function]}
          />
          <Hand
                    handType="rules"
          />
</div>,
        <WithCondition
          choseHandType={[Function]}
          handTypes={
                    Object {
                              "paper": "paper",
                              "rock": "rock",
                              "scissors": "scissors",
                            }
          }
          shouldShow={false}
/>,
        "}",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "cpuScores": 0,
          "playerScores": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Hand
              handType="rules"
              isCpu={false}
/>,
            <Desk
              history={Array []}
              isPlaying={false}
              playFunc={[Function]}
/>,
            <Hand
              handType="rules"
/>,
          ],
          "className": "game-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handType": "rules",
              "isCpu": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "history": Array [],
              "isPlaying": false,
              "playFunc": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handType": "rules",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "choseHandType": [Function],
          "handTypes": Object {
            "paper": "paper",
            "rock": "rock",
            "scissors": "scissors",
          },
          "shouldShow": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "}",
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ScoreBar
            cpuScores={0}
            playerScores={0}
/>,
          <div
            className="game-container"
>
            <Hand
                        handType="rules"
                        isCpu={false}
            />
            <Desk
                        history={Array []}
                        isPlaying={false}
                        playFunc={[Function]}
            />
            <Hand
                        handType="rules"
            />
</div>,
          <WithCondition
            choseHandType={[Function]}
            handTypes={
                        Object {
                                    "paper": "paper",
                                    "rock": "rock",
                                    "scissors": "scissors",
                                  }
            }
            shouldShow={false}
/>,
          "}",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "cpuScores": 0,
            "playerScores": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Hand
                handType="rules"
                isCpu={false}
/>,
              <Desk
                history={Array []}
                isPlaying={false}
                playFunc={[Function]}
/>,
              <Hand
                handType="rules"
/>,
            ],
            "className": "game-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handType": "rules",
                "isCpu": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "history": Array [],
                "isPlaying": false,
                "playFunc": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handType": "rules",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "choseHandType": [Function],
            "handTypes": Object {
              "paper": "paper",
              "rock": "rock",
              "scissors": "scissors",
            },
            "shouldShow": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "}",
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
